### ISSUE TITLE
Random permission denied when read approle secret-id-accessor

### ISSUE THREAD

**Issue by yura-shutkin** [hashicorp/vault#6568]
**Describe the bug**

Role with wildcard policy randomly can't "read" approle secret-id-accessor
```
I can't tell why and how. But I can't reproduce this every time. Maybe it may happen when laptop is overloaded.
Ayway: I will appreciate if you'll check this.
```

**To Reproduce**
* Preparation
1. Run `docker run --rm -ti -p 127.0.0.1:8200:8200 -e VAULT_DEV_ROOT_TOKEN_ID=12345 vault:1.1.0`
2. Run `vault login 12345`
3. Run `vault auth enable userpass`
4. Run `vault auth enable approle`
5. Run `vault write auth/userpass/users/wildcard password=12345 policies=default,policies=default,approle-supervisor-wildcard`
6. Run `vault write auth/userpass/users/hardcode password=12345 policies=default,policies=default,approle-supervisor-hardcode`
7. Run `vault write auth/approle/role/my-app policies=default`
8. Create file `hardcode.hcl` with
```
path "auth/approle/role" {
  capabilities = ["list"]
}

path "auth/approle/role/my-app/secret-id" {
  capabilities = ["list"]
}

path "auth/approle/role/my-app/secret-id-accessor/lookup" {
  capabilities = ["update"]
}
```
9. Create file `wildcard.hcl` with
```
path "auth/approle/role" {
  capabilities = ["list"]
}

path "auth/approle/role/+/secret-id" {
  capabilities = ["list"]
}

path "auth/approle/role/+/secret-id-accessor/lookup" {
  capabilities = ["update"]
}
```
10. Create approle-supervisor-hardcode policy `vault policy write approle-supervisor-hardcode hardcode.hcl`
11. Create approle-supervisor-wildcard policy `vault policy write approle-supervisor-wildcard wildcard.hcl`
12. Run `vault write -force auth/approle/role/my-app/secret-id`
13. Run `vault write -force auth/approle/role/my-app/secret-id`
14. Run `SECRET_ID_ACCESSOR=$(vault write -force -field=secret_id_accessor auth/approle/role/my-app/secret-id)`
15. Run `vault list auth/approle/role/my-app/secret-id`
* Hardcoded policy
1. Run `vault login -method=userpass username=hardcoded password=12345`
2. Run `for idx in {1..20}; do curl -sk -X POST -d "{\"secret_id_accessor\": \"$SECRET_ID_ACCESSOR\"}" -H "X-Vault-Token: $(cat ~/.vault-token)" $VAULT_ADDR/v1/auth/approle/role/my-app/secret-id-accessor/lookup; sleep 0.1; done`
3. See 
```{"request_id":"fc67ddbe-5e6f-5ce3-5ebb-b5d85dde1f65","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"0353a18d-c5e2-4f93-409d-0270ede3e86c","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"748d44c3-bd18-b8f5-2dff-bcd4fbbb4aaf","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"1dc8ea9e-d018-dfcb-7562-bb5b0d98ea93","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"3714e7f3-49a2-43e1-a7b6-1f681cba4b2f","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"9e3fb105-628f-738c-4a3c-3ceec62a2da4","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"b9fb4bc5-df41-e612-cee6-5dbee49f429b","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"a5f43e34-996b-5e0c-31fe-60170a70dcef","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"262f6402-c10a-bdc6-52ec-e7873f4604ea","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"86f46742-d614-cdf1-fa51-399f429a7b26","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"ff8dce5d-dad5-fa6e-d518-0a573dc35a77","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"6ab66f03-9eb1-97e9-d737-20c7b241d0f7","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"bb120473-e78e-a05b-825e-7693746404bc","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"d9f41fec-227e-7bdc-b285-97cefe4557d8","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"4bf77d26-c675-a27a-e806-39502f82c79b","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"63da49d1-b00f-043d-b9ba-a70c5a3939ab","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"6d118a46-9889-768d-db10-cbebe437a5dd","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"dd18a17c-4f73-7191-f6f8-e9cffdb06752","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"cbfadb84-b35f-7d8f-41fc-b4af9d7e872e","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"request_id":"8fc533ef-c898-6926-4437-88d7dc096d70","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
```
* Wildcard policy
1. Run `vault login -method=userpass username=hardcoded password=12345`
2. Run `for idx in {1..20}; do curl -sk -X POST -d "{\"secret_id_accessor\": \"$SECRET_ID_ACCESSOR\"}" -H "X-Vault-Token: $(cat ~/.vault-token)" $VAULT_ADDR/v1/auth/approle/role/my-app/secret-id-accessor/lookup; sleep 0.1; done`
3. See 
```{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"request_id":"bccb60cf-eea5-b813-a3bf-583cb79f1b7b","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"request_id":"c3ead49c-930b-9094-784c-a62dc7e2ebd8","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"request_id":"22838986-f8b8-0f6d-c106-5c6410177880","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"request_id":"ff5b825f-4e6f-dab8-3fda-281ded37cd17","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"errors":["1 error occurred:\n\t* permission denied\n\n"]}
{"request_id":"ab4355b2-b3db-f234-ac25-fce6562bb6fa","lease_id":"","renewable":false,"lease_duration":0,"data":{"cidr_list":[],"creation_time":"2019-04-11T08:12:45.844627712Z","expiration_time":"0001-01-01T00:00:00Z","last_updated_time":"2019-04-11T08:12:45.844627712Z","metadata":{},"secret_id_accessor":"f171d13e-8af9-b105-46b5-eea1eca822d4","secret_id_num_uses":0,"secret_id_ttl":0,"token_bound_cidrs":[]},"wrap_info":null,"warnings":null,"auth":null}
```

**Expected behavior**

Policy with wildcard path should allow "read" secret-id-accessor as hardcoded does

**Environment:**
* Vault Server Version (retrieve with `vault status`): 
```
Key             Value
---             -----
Seal Type       shamir
Initialized     true
Sealed          false
Total Shares    1
Threshold       1
Version         1.1.0
Cluster Name    vault-cluster-b5ba6bb3
Cluster ID      6eaad36c-e37e-1d29-fb38-862fc63beff0
HA Enabled      false
```
* Vault CLI Version (retrieve with `vault version`):
`Vault v1.1.0 ('36aa8c8dd1936e10ebd7a4c1d412ae0e6f7900bd')`
* Server Operating System/Architecture:
```
Docker version 18.09.4, build d14af54266 on Ubuntu 18.04
```

Vault server configuration file(s):
`dev mode no config required`

**Additional context**

script for automation
```bash
#!/usr/bin/env bash

export VAULT_ADDR=http://127.0.0.1:8200
vault login 12345
vault auth enable userpass
vault auth enable approle
vault write auth/userpass/users/wildcard password=12345 policies=default,policies=default,approle-supervisor-wildcard
vault write auth/userpass/users/hardcode password=12345 policies=default,policies=default,approle-supervisor-hardcode
vault write auth/approle/role/my-app policies=default
echo 'path "auth/approle/role" {
  capabilities = ["list"]
}

path "auth/approle/role/my-app/secret-id" {
  capabilities = ["list"]
}

path "auth/approle/role/my-app/secret-id-accessor/lookup" {
  capabilities = ["update"]
}' > hardcode.hcl
echo 'path "auth/approle/role" {
  capabilities = ["list"]
}

path "auth/approle/role/+/secret-id" {
  capabilities = ["list"]
}

path "auth/approle/role/+/secret-id-accessor/lookup" {
  capabilities = ["update"]
}' > wildcard.hcl
vault policy write approle-supervisor-hardcode hardcode.hcl
vault policy write approle-supervisor-wildcard wildcard.hcl
vault write -force auth/approle/role/my-app/secret-id
vault write -force auth/approle/role/my-app/secret-id
SECRET_ID_ACCESSOR=$(vault write -force -field=secret_id_accessor auth/approle/role/my-app/secret-id)
vault list auth/approle/role/my-app/secret-id
# hardcode
vault login -method=userpass username=hardcode password=12345
for idx in {1..20}; do echo $idx;  curl -sk -X POST -d "{\"secret_id_accessor\": \"$SECRET_ID_ACCESSOR\"}" -H "X-Vault-Token: $(cat ~/.vault-token)" $VAULT_ADDR/v1/auth/approle/role/my-app/secret-id-accessor/lookup; sleep 0.1; done
# wildcard
vault login -method=userpass username=wildcard password=12345
for idx in {1..20}; do curl -sk -X POST -d "{\"secret_id_accessor\": \"$SECRET_ID_ACCESSOR\"}" -H "X-Vault-Token: $(cat ~/.vault-token)" $VAULT_ADDR/v1/auth/approle/role/my-app/secret-id-accessor/lookup; sleep 0.1; done

```

**Comment by ncabatoff** [comment]
Hi @yura-shutkin , 

Thank you for the detailed bug report and steps to reproduce.

I've looked closely at the 1.1.0 policy code and I can't see what might make this happen.  I've tried to reproduce it with https://github.com/ncabatoff/vault-experiments/blob/master/bugrepros/gh5658.sh to no avail.  I also don't see any way in which load could be a factor here.

I'm out of ideas, sorry.  Do let us know if you think of any other details that might be relevant.
